### ISSUE TITLE
Idea: vectorized environments (parallel games)

### ISSUE THREAD

**Issue by eriiiko** [Apelsinmack/Civulator#22]
Run 8+ games in parallel per training step (NEXT_STEPS.md ML roadmap). More viable as more of the game core moves to C++ (`civulator_core`).

**Comment by eriiiko** [comment]
Upgraded from idea to motivated-by-data (Erik, 2026-08-23: 'batching looks quite appealing, if we can speed up we should'). #34 profiling found the training night 79% torch-bound with a kernel-launch-bound smell (694k tiny conv/batchnorm calls on a 12×24 board) — batching is the lever that reaches it, and vectorized envs are the natural batching vehicle. IMPORTANT scientific caveat for whoever picks this up: unlike the #42 mask fix (bit-identical, neutral), batching changes are NOT automatically neutral — batch size and parallel data collection change training dynamics, i.e., they are experiment variables. Sequence: (1) torch.profiler kernel trace to confirm launch-bound (pure measurement); (2) semantics-preserving engineering wins first (fusion, CUDA graphs, fewer H2D transfers); (3) vectorized envs / batch-size changes as a MEASURED experiment against the #39 baseline, never a silent speedup mid-series.

**Comment by eriiiko** [comment]
## Design doc written: `docs/training_throughput_design.md`

Erik + Claude, 2026-09-05. Lit search done (EnvPool, Gymnasium, Parallel Q-Learning, Policy Churn, AcceRL); sources listed in the doc. **Slot agreed: after the anti-turtling ladder's last run completes, before #71.**

### The finding that changes this issue's premise

`learning_step` — the 78.8% from #34 — is `DQNAgent.optimize`, i.e. the **learner**, not the actor. It runs **one gradient step per agent action at batch 32** (`trainer.py:325-326`): a replay ratio of 1.0 updates/transition. Splitting the per-action cost into `L = 0.788` (the gradient step; strictly sequential, since update *k+1* reads the weights update *k* wrote) and `A = 0.212` (everything else, batchable across games):

```
N parallel games, replay ratio held fixed:
  wall-clock per game ≈ L·T + A·T/N
  speedup = 1 / (L + A/N)  →  1.22× at N=8,  1.27× as N→∞
```

**Vectorized environments alone cannot exceed ~1.27×, at any N.** Amdahl, with the learner as the serial fraction. And the engine is 1.2% of the night (pathfinding 0.17%), so there is no simulation headroom to reclaim either — the observation that "the CPU is far from 100% during training" is the signature of a serialized single-threaded loop on a 16-thread box, not of spare capacity.

The gain lives entirely in what vectorization *enables*: one big-batch gradient step per N transitions instead of N small ones, i.e. replay ratio 1.0 → 1/N. **The speedup and the scientific confound are therefore the same change** — which is why the doc's §9 protocol varies replay ratio alone at N=1 first, as the control, before varying N at all. This sharpens rather than contradicts the caveat already in this thread.

### Two things that came out of it

**A neutral win worth more than the vectorization.** `compute_loss` runs the target network **twice on the same `next_state_batch`** (`dqn_agent.py:248,251`), and at batch 32 on a 12×24 board the whole graph is almost certainly launch-bound. Fusing that + CUDA graphs on `optimize()` is bit-for-bit the same math and plausibly ~2.4× on the whole night — no experiment, no confound. Landable independently of anything here.

**Civ's sequentiality decides the architecture, not just the feasibility.** There is no batch axis inside a game (turn *t+1* depends on *t*; within a turn, unit A's move changes unit B's legal moves). The batch axis is games and only games — and games desynchronise immediately, so there is no lockstep step index to batch on. That is exactly why AcceRL's **dynamic windowing** (batch whatever requests are pending, keyed on `(seat, network)`) is the right pattern here and a lockstep vector env is not. Its inference-as-a-service framing also decouples N from the replay ratio, turning lever B into a controlled experiment. GPU-side simulation stays rejected per #32 — nothing here reopens it.

### One measurement gates the whole plan

Time `optimize()` at batch 32 → 1024, CUDA-synced, warm, **on an idle GPU**:

- **Flat curve** → launch-bound → the plan above holds.
- **Linear curve** → compute-bound → a batch-32N step costs N× a batch-32 step, **lever B collapses**, and the honest recommendation becomes "do the neutral fixes, skip parallel envs". The doc states this alternative explicitly so it is allowed to come back negative.

No parallel-env code is written before that reports. Patch sequence in §8, experimental protocol in §9, Systems inventory in §10.

**Caveat on the numbers**: #34's shares predate #42, #40's 52ch encoder and #48, and the in-flight run trains at 54ch — so a re-profile at current config is gate 0, not an assumption.

**Comment by eriiiko** [comment]
## Two follow-ups from the 2026-09-05 discussion

### 1. Where to set the replay ratio — two criteria that agree

Using `P(a transition is never sampled) = e^(−batch_size·r) = e^(−32r)` (uniform sampling, full buffer — see the design doc §4B):

- **1% never sampled** → `32r = ln(100)` → **r ≈ 0.144**, i.e. one update every ~7 actions. Speedup **3.07×**.
- **Where the speedup curve bends** → the learner stops dominating when `L·r = A`, i.e. `0.788r = 0.212` → **r ≈ 0.27**, one update every ~4 actions. Speedup **2.36×**. Below that, each further halving buys progressively less against the 4.72× ceiling.

**The two criteria bracket the same region: r ∈ [0.14, 0.27] — update every 4 to 7 actions.** That window contains both the classic DQN default (every 4, r = 0.25) and Erik's independent guess of every 5 (r = 0.20 → 3.0× speedup, 0.17% never sampled). The P5/P6 ladder should therefore concentrate its samples in this range rather than spanning orders of magnitude.

Caveat for later: ranking transitions (prioritized experience replay, which Erik wants to explore) **invalidates this formula** — it assumes uniform sampling.

### 2. Evaluation is the vector env's risk-free first customer

Evaluation has **no learner**, so `L = 0` and the entire cost is `A`. That inverts the arithmetic that made vectorization unattractive for training:

```
training at r=1:  speedup = 1/(0.788 + 0.212/N)  →  1.22× at N=8
evaluation:       speedup = A/(A/N)              →  ≈ N
```

Near-linear speedup, and — the part that matters — **zero scientific confound**: no replay ratio, no training dynamics, no experiment variable. Same weights, same seeds, same games, just more in flight. Correctness is directly checkable: the same seeds must yield the same win counts as the sequential path.

So P7 (`VectorGameEnvironment`) should be built **against `scripts/evaluate.py` first**, where it can be validated, before it ever touches the trainer.

*Caveat*: batched convolutions may select different cuDNN algorithms than batch-1, so results may not be strictly bit-identical — a flipped argmax on a near-tie could change a game. The acceptance test is "same seeds → same outcomes", with any divergence investigated (this is where #69's state fingerprint earns its keep) rather than waved through.

### 3. A cheaper immediate option for slow evaluations

`scripts/evaluate.py:351` is a plain sequential `for i in range(games)` — embarrassingly parallel. Sharding the seed range across K processes and merging the JSONs needs no new architecture. Two things to get right:

- the paired-sides protocol means each seed's A/B pair must stay **in the same shard**;
- the merge must carry `truncated_games` / `games_detail[]` through intact (#51).

Worth measuring first whether evaluation is even GPU-worthy: a single-sample forward through a small FullyConv net is exactly the regime where launch overhead exceeds compute, so K CPU processes may beat one GPU process with no VRAM ceiling. **Note #81 currently blocks the CPU-side version of this** — the device idiom resolves to `cuda` whenever devices are masked.

**Comment by eriiiko** [comment]
**Gate 1 report — batch-scaling microbenchmark** (`scripts/bench_learning_step.py`, merged as defe782; artifact `stats/bench_learning_step_1788732617.json`, manifest-embedded, git_commit d9f50b6, RTX 3070, torch 2.6.0+cu124, duel 24×12, encoder `enhanced` 25ch, conv_channels (16,32), 20 warmup + 100 timed calls per point).

| batch | ms/call (mean±std) | transitions/s |
|---:|---:|---:|
| 32 | 13.29 ± 1.23 | 2409 |
| 64 | 22.40 ± 2.76 | 2858 |
| 128 | 33.71 ± 2.16 | 3797 |
| 256 | 62.19 ± 3.34 | 4117 |
| 512 | 132.43 ± 15.11 | 3866 |
| 1024 | 242.12 ± 19.15 | 4229 |

**ms(1024)/ms(32) = 18.2×** against a 32× batch increase — neither flat nor cleanly linear. A linear fit gives a fixed component of ≈5.9 ms/call (≈44% of a batch-32 call) and ≈0.23 ms per sample; transitions/s rises only 1.76× from batch 32→1024.

**Verdict for the plan**: mostly compute-bound at this architecture. Lever A (P2 fused double target-forward, P3 H2D staging, P4 CUDA graphs) targets the ≈5.9 ms fixed cost and stays worth doing. Lever B's derived N× gains are an **upper bound, not an expectation** — Gate 2's kernel trace should split the fixed cost into launch overhead vs small fixed compute (target-net sync, Adam overhead) before P4 is sized.

**Caveats**: (a) config.toml pins no conv_channels; the bench used the trainer/run_baseline pinned (16,32) — the strongest manifest models are net128x6, where the shape would shift further toward compute-bound; worth a second data point before betting on lever B. (b) On this Windows box `nvidia-smi --query-compute-apps` lists idle desktop apps — the script's idle guard uses `utilization.gpu` (>5% aborts) instead, and tripped correctly once during testing. (c) One timed call in 100 can naturally include the target_update_freq=100 state_dict copy — left in as production-faithful; plausible source of the batch-512 stdev.

Measured by an autonomous run 2026-09-07; suite 675 passed / 2 skipped on the branch. Gate 0 (re-profile + steps/turns/optimize_count instrumentation) runs next.
